### DISCUSSION TITLE
Problems with controller events in release mode

### DISCUSSION THREAD

**Discussion by paulocoutinhox** [axmolengine/axmol#1986] [General]
Hi,

Im testing my game on AppleTV in debug/release (debug and release from xcode and release from testflight), but i have a problem with GameController in release mode.

In debug it works perfect, but if i run in release mode, the events don't come.

Anyone knows what can be wrong?

I have this test code that is a fairygui window:

```cpp
void MenuWindowGUI::onInit() {
    // create and set content pane
    auto contentPane = UIPackage::createObject("Main", "GameMenuWindow")->as<GComponent>();
    setContentPane(contentPane);
    EnvironmentHelper::centerContentPane(contentPane);

    // initialize control list
    controlList.push_back("btStart1");
    controlList.push_back("btStart2");
    controlList.push_back("btStart3");
    controlListIndex = 0;

    // add other control names as needed
    // example: controlList.push_back("btOptions");

    // select the first control by default
    updateControlSelection();

    // add click listeners for buttons
    contentPane->getChildByPath("btStart1")->addClickListener(CC_CALLBACK_1(MenuWindowGUI::onBtStartClick, this));
    contentPane->getChildByPath("btStart2")->addClickListener(CC_CALLBACK_1(MenuWindowGUI::onBtStartClick, this));
    contentPane->getChildByPath("btStart3")->addClickListener(CC_CALLBACK_1(MenuWindowGUI::onBtStartClick, this));
    // add other button click listeners as needed

    // add joystick listener
    Controller::startDiscoveryController();
    auto listener = EventListenerController::create();
    listener->onKeyDown = AX_CALLBACK_3(MenuWindowGUI::onControllerKeyDown, this);
    listener->onKeyUp = AX_CALLBACK_3(MenuWindowGUI::onControllerKeyUp, this);

    // get the current scene node
    auto scene = Director::getInstance()->getRunningScene();
    if (scene) {
        scene->getEventDispatcher()->addEventListenerWithSceneGraphPriority(listener, scene);
    }
}
```

Thanks.

**Comment by paulocoutinhox** [comment]
How it is not working in release mode, i put some error log to force show messages in console since in release it only show error messages, and the only message in console is the first `AXLOGE("EventListenerController::init");`. The others when event is received, don't happen in release mode.

```cpp
bool EventListenerController::init()
{
    AXLOGE("EventListenerController::init");
    
    auto listener = [this](Event* event) {
        auto evtController = static_cast<EventController*>(event);
        switch (evtController->getControllerEventType())
        {
        case EventController::ControllerEventType::CONNECTION:
            AXLOGE("EventController::ControllerEventType::CONNECTION - {}", evtController->isConnected());
                
            if (evtController->isConnected())
            {
                if (this->onConnected)
                    this->onConnected(evtController->getController(), event);
            }
            else
            {
                if (this->onDisconnected)
                    this->onDisconnected(evtController->getController(), event);
            }
            break;
        case EventController::ControllerEventType::BUTTON_STATUS_CHANGED:
        {
            AXLOGE("EventController::ControllerEventType::BUTTON_STATUS_CHANGED - {}", evtController->_keyCode);
            
            const auto& keyStatus     = evtController->_controller->_allKeyStatus[evtController->_keyCode];
            const auto& keyPrevStatus = evtController->_controller->_allKeyPrevStatus[evtController->_keyCode];

            if (this->onKeyDown && keyStatus.isPressed && !keyPrevStatus.isPressed)
            {
                this->onKeyDown(evtController->_controller, evtController->_keyCode, event);
            }
            else if (this->onKeyUp && !keyStatus.isPressed && keyPrevStatus.isPressed)
            {
                this->onKeyUp(evtController->_controller, evtController->_keyCode, event);
            }
            else if (this->onKeyRepeat && keyStatus.isPressed && keyPrevStatus.isPressed)
            {
                this->onKeyRepeat(evtController->_controller, evtController->_keyCode, event);
            }
        }
        break;
        case EventController::ControllerEventType::AXIS_STATUS_CHANGED:
        {
            if (this->onAxisEvent)
            {
                this->onAxisEvent(evtController->_controller, evtController->_keyCode, event);
            }
        }
        break;
        default:
            AXASSERT(false, "Invalid EventController type");
            break;
        }
    };

    if (EventListener::init(EventListener::Type::GAME_CONTROLLER, LISTENER_ID, listener))
    {
        return true;
    }
    return false;
}
```

**Reply by rh101** [reply]
> How it is not working in release mode, i put some error log to force show messages in console since in release it only show error messages, and the only message in console is the first `AXLOGE("EventListenerController::init");`. The others when event is received, don't happen in release mode.

The logger can be configured to show any level of messages, not just error messages.  The function is `setLogLevel`, which is located in the `Logger.h/.cpp` files.

**Comment by thienphuoc** [comment]
To verify whether the issue is caused by compiler optimizations in release mode, you should try changing the Optimization Level for the release configuration to None [-O0]. 

This will disable all compiler optimizations and help you determine if the optimizations are the source of the problem.

![Screenshot 2024-06-12 at 14 52 20](https://github.com/axmolengine/axmol/assets/17909394/f90ab960-cc91-4566-a4fa-57c565f2831e)

**Reply by paulocoutinhox** [reply]
I made, but it don't work, same problem:

<img width="1512" alt="image" src="https://github.com/axmolengine/axmol/assets/395096/8c7f97f1-3a9b-45c6-a2b3-029b2bb006a2">

**Reply by paulocoutinhox** [reply]
Need set to all targets?

**Reply by thienphuoc** [reply]
EventListenerController is located in core/base. I think you should set the target to axmol. 

![image](https://github.com/axmolengine/axmol/assets/17909394/03083019-760f-4e29-b57a-e736fa0f2d29)

**Reply by paulocoutinhox** [reply]
I set to all, but the problem still happen.

**Comment by paulocoutinhox** [comment]
I think that there is a bug in Axmol @rh101 because when i set the macro `_AX_DEBUG = 1` it works without problems in release mode.

<img width="864" alt="image" src="https://github.com/axmolengine/axmol/assets/395096/6fa06945-b415-4299-9338-942b2f0db378">

**Reply by paulocoutinhox** [reply]
I even tried to see if there was anything obvious that wasn't being called with the _AX_DEBUG macro but I didn't find anything obvious that could be directly related to this.

**Reply by rh101** [reply]
> I think that there is a bug in Axmol @rh101 because when i set the macro `_AX_DEBUG = 1` it works without problems in release mode.

Sorry, I don't understand what you mean by that.  What works? Are you referring to the logging functions, or to the controller events?

**Reply by paulocoutinhox** [reply]
The main problem, the controller events. When i set `_AX_DEBUG=1` it start works in release mode. Can you help me?

**Reply by rh101** [reply]
> The main problem, the controller events. When i set `_AX_DEBUG=1` it start works in release mode. Can you help me?

That is strange, and doesn't make much sense at all.  Have you reviewed your code to check if you have accidentally surrounded code by conditional checks for debug mode?

Is there a link to the project source code available?

**Reply by AlexandreK38** [reply]
> I think that there is a bug in Axmol @rh101 because when i set the macro _AX_DEBUG = 1 it works without problems in release mode.

But setting `_AX_DEBUG=1` in the `AppDelegate` will only affect your app code, not the axmol library built without it, as the axmol lib has this definition in the Preprocessor Macro for which the debug is enabled (`Debug` and `RelWithDebInfo`)
To me, it would mean it's your app code acting differently according to the _AX_DEBUG

**Comment by paulocoutinhox** [comment]
I still don't know what is wrong. I created a new project and put the same code, a scene with fairygui and listener for events and it is working.

https://github.com/paulocoutinhox/axmol-controller-test

**Comment by paulocoutinhox** [comment]
Hi,

I solve the problem and this is the code changes:
https://github.com/paulocoutinhox/axmol-controller-test/commit/5be30721f3494df38ba39d930b5ed85e374d26d6

Basically, in release it don't work (don't ask me why):

```
// add joystick listener
Controller::startDiscoveryController();
auto listener = EventListenerController::create();
listener->onKeyDown = AX_CALLBACK_3(MenuWindowGUI::onControllerKeyDown, this);

// get the current scene node
auto scene = Director::getInstance()->getRunningScene();
if (scene) {
    scene->getEventDispatcher()->addEventListenerWithSceneGraphPriority(listener, scene);
}
```

What i do is get the event from scene and send to UI class.
